catalog/sql/sql.go

```diff
@@ -0,0 +1,721 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package sql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"maps"
+	"path"
+	"slices"
+	"strings"
+	"sync"
+	_ "unsafe"
+
+	"github.com/apache/iceberg-go"
+	"github.com/apache/iceberg-go/catalog"
+	"github.com/apache/iceberg-go/catalog/internal"
+	"github.com/apache/iceberg-go/io"
+	"github.com/apache/iceberg-go/table"
+	"github.com/uptrace/bun"
+	"github.com/uptrace/bun/dialect/feature"
+	"github.com/uptrace/bun/dialect/mssqldialect"
+	"github.com/uptrace/bun/dialect/mysqldialect"
+	"github.com/uptrace/bun/dialect/oracledialect"
+	"github.com/uptrace/bun/dialect/pgdialect"
+	"github.com/uptrace/bun/dialect/sqlitedialect"
+	"github.com/uptrace/bun/extra/bundebug"
+	"github.com/uptrace/bun/schema"
+)
+
+type SupportedDialect string
+
+const (
+	Postgres SupportedDialect = "postgres"
+	MySQL    SupportedDialect = "mysql"
+	SQLite   SupportedDialect = "sqlite"
+	MSSQL    SupportedDialect = "mssql"
+	Oracle   SupportedDialect = "oracle"
+)
+
+const (
+	DialectKey           = "sql.dialect"
+	DriverKey            = "sql.driver"
+	initCatalogTablesKey = "init_catalog_tables"
+)
+
+func init() {
+	catalog.Register("sql", catalog.RegistrarFunc(func(name string, p iceberg.Properties) (c catalog.Catalog, err error) {
+		driver, ok := p[DriverKey]
+		if !ok {
+			return nil, errors.New("must provide driver to pass to sql.Open")
+		}
+
+		dialect := strings.ToLower(p[DialectKey])
+		if dialect == "" {
+			return nil, errors.New("must provide sql dialect to use")
+		}
+
+		uri := strings.TrimPrefix(p.Get("uri", ""), "sql://")
+		sqldb, err := sql.Open(driver, uri)
+		if err != nil {
+			return nil, err
+		}
+
+		defer func() {
+			if r := recover(); r != nil {
+				err = fmt.Errorf("failed to create SQL catalog: %v", r)
+			}
+		}()
+
+		return NewCatalog(p.Get(name, "sql"), sqldb, SupportedDialect(dialect), p)
+	}))
+}
+
+var (
+	_ catalog.Catalog = (*Catalog)(nil)
+)
+
+var (
+	minimalNamespaceProps = iceberg.Properties{"exists": "true"}
+
+	dialects  = map[SupportedDialect]schema.Dialect{}
+	dialectMx sync.Mutex
+)
+
+func createDialect(d SupportedDialect) schema.Dialect {
+	switch d {
+	case Postgres:
+		return pgdialect.New()
+	case MySQL:
+		return mysqldialect.New()
+	case SQLite:
+		return sqlitedialect.New()
+	case MSSQL:
+		return mssqldialect.New()
+	case Oracle:
+		return oracledialect.New()
+	default:
+		panic("unsupported sql dialect")
+	}
+}
+
+func getDialect(d SupportedDialect) schema.Dialect {
+	dialectMx.Lock()
+	defer dialectMx.Unlock()
+	ret, ok := dialects[d]
+	if !ok {
+		ret = createDialect(d)
+		dialects[d] = ret
+	}
+	return ret
+}
+
+type sqlIcebergTable struct {
+	bun.BaseModel `bun:"table:iceberg_tables"`
+
+	CatalogName              string `bun:",pk"`
```

Comment:
Something I want to point out when including the `CatalogName` in the underlying DB. 
This name is passed to the catalog at runtime and persisted in the underlying DB. When it is access again, the same name must be provided to get the same namespace/tables back.

This might be a somewhat odd behavior because the "name" of the catalog doesn't/shouldn't hold any semantic meanings. 

For something like REST catalog, I can name it "foo" the first time and "bar" the second time, as long as its pointing to the same URL, i get the same tables back.

Reply:
At the moment I was following the pattern in pyiceberg and JDBC so that everything is compatible. While I agree with you that the "name" shouldn't hold any semantic meaning, I'd prefer to stay consistent with the other implementations for now, unless you think we should update the other implementations too? :) haha